File: bench/suite/src/st/spectral-norm/javascript/spectral-norm.js
// spectral-norm: single-threaded, plain JavaScript (Benchmarks Game algorithm).
'use strict';
function A(i, j) { return 1 / (((i + j) * (i + j + 1) >>> 1) + i + 1); }
function Au(u, v, n) {
  for (let i = 0; i < n; i++) {
    let s = 0;
    for (let j = 0; j < n; j++) s += A(i, j) * u[j];
    v[i] = s;
  }
}
function Atu(u, v, n) {
  for (let i = 0; i < n; i++) {
    let s = 0;
    for (let j = 0; j < n; j++) s += A(j, i) * u[j];
    v[i] = s;
  }
}
function AtAu(u, v, w, n) { Au(u, w, n); Atu(w, v, n); }
const n = +process.argv[2] || 100;
const u = new Float64Array(n).fill(1), v = new Float64Array(n), w = new Float64Array(n);
for (let i = 0; i < 10; i++) { AtAu(u, v, w, n); AtAu(v, u, w, n); }
let vBv = 0, vv = 0;
for (let i = 0; i < n; i++) { vBv += u[i] * v[i]; vv += v[i] * v[i]; }
console.log(Math.sqrt(vBv / vv).toFixed(9));

File: bench/suite/src/st/spectral-norm/python/spectral-norm.py
# spectral-norm: single-threaded, pure CPython (Benchmarks Game algorithm).
# Plain accumulation loops (not sum(), which is compensated since 3.12) so
# the arithmetic matches the reference program.
import sys
from math import sqrt


def mul_av(u, v, n):
    for i in range(n):
        s = 0.0
        for j in range(n):
            s += 1.0 / ((i + j) * (i + j + 1) // 2 + i + 1) * u[j]
        v[i] = s


def mul_atv(u, v, n):
    for i in range(n):
        s = 0.0
        for j in range(n):
            s += 1.0 / ((i + j) * (i + j + 1) // 2 + j + 1) * u[j]
        v[i] = s


def mul_atav(u, v, w, n):
    mul_av(u, w, n)
    mul_atv(w, v, n)


def main():
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 100
    u = [1.0] * n
    v = [0.0] * n
    w = [0.0] * n
    for _ in range(10):
        mul_atav(u, v, w, n)
        mul_atav(v, u, w, n)
    vbv = 0.0
    vv = 0.0
    for i in range(n):
        vbv += u[i] * v[i]
        vv += v[i] * v[i]
    print("%0.9f" % sqrt(vbv / vv))


main()
